package/MDAnalysis/lib/nsgrid.pyx

```diff
@@ -0,0 +1,946 @@
+            for i in range(ncoords):
+
+                # Add bead to grid cell
+                cellindex = self.cellids[i]
+                self.beadids[cellindex * self.nbeads_per_cell + beadcounts[cellindex]] = i
+                beadcounts[cellindex] += 1
+
+
+
+cdef class FastNS(object):
+    """Grid based search between two group of atoms
+
+    Instantiates a class object which uses :class:`PBCBox` and
+    :class:`NSGrid` to construct a cuboidal
+    grid in an orthogonal brick shaped box.
+
+    Minimum image convention is used for distance evaluations
+    if box dimensions are provided. 
+    """
+   
+    cdef PBCBox box
+    cdef real[:, ::1] coords
+    cdef real[:, ::1] coords_bbox
+    cdef readonly real cutoff
+    cdef NSGrid grid
+    cdef ns_int max_gridsize
+    cdef bint periodic
+    
+
+    def __init__(self, cutoff, coords, box=None, max_gridsize=5000):
+        """
+        Initialize the grid and sort the coordinates in respective
+        cells by shifting the coordinates in a brick shaped box.
+        The brick shaped box is defined by :class:`PBCBox`
+        and cuboidal grid is initialize by :class:`NSGrid`.
+        If box is supplied, periodic shifts along box vectors are used
+        to contain all the coordinates inside the brick shaped box.
+        If box is not supplied, the range of coordinates i.e.
+        ``[xmax, ymax, zmax] - [xmin, ymin, zmin]`` is used to construct
+        a pseudo box. Subsequently, the origin is also shifted
+        to the ``[xmin, ymin, zmin]``.
+
+        Parameters
+        ----------
+        cutoff : float
+            Desired cutoff distance 
+        coords : numpy.ndarray
+            atom coordinates of shape ``(N, 3)`` for ``N`` atoms.
+            ``dtype=numpy.float32``
+        box : numpy.ndarray
+            Box dimension of shape (6, ). The dimensions must be
+            provided in the same format as returned
+            by :attr:`MDAnalysis.coordinates.base.Timestep.dimensions`:
+            ``[lx, ly, lz, alpha, beta, gamma]`` (dtype = numpy.float32)
+            [None]
+        max_gridsize : int 
+            maximum number of cells in the grid
+
+        Note
+        ----
+        box=``None`` and ``[0., 0., 0., 90., 90., 90]``
+        both are used for Non-PBC aware calculations
+
+        """
+        import MDAnalysis as mda
+        from MDAnalysis.lib.mdamath import triclinic_vectors
+
+        cdef real[:] pseudobox = np.zeros(6, dtype=np.float32)
+        cdef real[DIM] bmax, bmin
+        cdef ns_int i
+        self.periodic = True
+
+
+        if (box is None) or (np.allclose(box[:3], 0.) and box.shape[0] == 6):
+            bmax = np.max(coords, axis=0)
+            bmin = np.min(coords, axis=0)
+            for i in range(DIM):
+                pseudobox[i] = 1.1*(bmax[i] - bmin[i])
+                pseudobox[DIM + i] = 90.
+            box = pseudobox
+            # shift the origin
+            coords -= bmin

```

Comment:
Here you're modifying coords in place, which changes the copy that the user holds (which might lead to tears).  Better would be to move the `self.coords = coords.copy()` call below, and modify `self.coords` which is the class' copy.